gslib/iamcredentials_api.py

```diff
@@ -0,0 +1,258 @@
+    """Generates an access token for the given service account."""
+    name = 'projects/-/serviceAccounts/%s' % service_account_id
+    generate_access_token_request = apitools_messages.GenerateAccessTokenRequest(
+        scope=scopes)
+    request = (apitools_messages.
+               IamcredentialsProjectsServiceAccountsGenerateAccessTokenRequest(
+                   name=name,
+                   generateAccessTokenRequest=generate_access_token_request))
+
+    try:
+      return self.api_client.projects_serviceAccounts.GenerateAccessToken(
+          request)
+    except TRANSLATABLE_APITOOLS_EXCEPTIONS as e:
+      raise self._TranslateExceptionAndRaise(e, service_account_id)
+
+  def _TranslateExceptionAndRaise(self, e, service_account_id=None):
+    """Translates an HTTP exception and raises the translated or original value.
+
+    Args:
+      e: Any Exception.
+      service_account_id: Optional service account in request that caused the exception.
+
+    Raises:
+      Translated CloudApi exception, or the original exception if it was not
+      translatable.
+    """
+    if self.logger.isEnabledFor(logging.DEBUG):
+      self.logger.debug('TranslateExceptionAndRaise: %s',
+                        traceback.format_exc())
+    translated_exception = self._TranslateApitoolsException(
+        e, service_account_id=service_account_id)
+    if translated_exception:
+      raise translated_exception
+    else:
+      raise
+
+  def _GetMessageFromHttpError(self, http_error):
+    if isinstance(http_error, apitools_exceptions.HttpError):
+      if getattr(http_error, 'content', None):
+        try:
+          json_obj = json.loads(http_error.content)
+          if 'error' in json_obj and 'message' in json_obj['error']:
+            return json_obj['error']['message']
+        except Exception:  # pylint: disable=broad-except
+          # If we couldn't decode anything, just leave the message as None.
+          pass
+
+  def _GetAcceptableScopesFromHttpError(self, http_error):
+    try:
+      www_authenticate = http_error.response['www-authenticate']
+      # In the event of a scope error, the www-authenticate field of the HTTP
+      # response should contain text of the form
+      #
+      # 'Bearer realm="https://oauth2.googleapis.com/",
+      # error=insufficient_scope,
+      # scope="${space separated list of acceptable scopes}"'
+      #
+      # Here we use a quick string search to find the scope list, just looking
+      # for a substring with the form 'scope="${scopes}"'.
+      scope_idx = www_authenticate.find('scope="')
+      if scope_idx >= 0:
+        scopes = www_authenticate[scope_idx:].split('"')[1]
+        return 'Acceptable scopes: %s' % scopes
+    except Exception:  # pylint: disable=broad-except
+      # Return None if we have any trouble parsing out the acceptable scopes.
+      pass
+
+  def _TranslateApitoolsException(self, e, service_account_id=None):
```

Comment:
Nit: I might move this up above _GetMessageFromHttpError.  I think the file reads better when you order the methods top down.